src/libplctag/Tag.cs

```diff
@@ -111,9 +115,83 @@ private static string GetAttributeString(Protocol protocol, IPAddress gateway, s
 
         public void Abort() => plctag.abort(pointer);
 
-        public void Read(int millisecondTimeout) => plctag.read(pointer, millisecondTimeout);
+        public void Read(int millisecondTimeout)
+        {
+
+            if (millisecondTimeout <= 0)
+                throw new ArgumentOutOfRangeException(nameof(millisecondTimeout), "Must be greater than 0 for a synchronous read");
+
+            var status = (Status)plctag.read(pointer, millisecondTimeout);
+            if (status == Status.Ok)
+                return;
+            else
+                throw new LibPlcTagException(status);
+
+        }
+
+        public async Task ReadAsync(CancellationToken token = default)
+        {
+
+            var status = (Status)plctag.read(pointer, 0);
+
+            if (status == Status.Ok)
+                return;
+            else if (status < 0)
+                throw new LibPlcTagException(status);
+
+            do
+            {
+                token.ThrowIfCancellationRequested();
```

Comment:
Do we want to throw the task cancelled exception to the caller? I could go either way. If you call an async and then cancel it, I think it's okay for it to just return and not throw an exception. In this case we could try/catch the cancellation exception internally or we could add `toke.IsCancelled` to the while condition.